segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/ThrottlerTests.java

```diff
@@ -145,6 +161,191 @@ public void testInterruptedDurableDataLogDelay() throws Exception {
+        val cacheCalculator = new TestCalculatorThrottler(ThrottlerCalculator.ThrottlerName.Cache);
+        val batchingCalculator = new TestCalculatorThrottler(ThrottlerCalculator.ThrottlerName.Batching);
+        val durableDataLogCalculator = new TestCalculatorThrottler(ThrottlerCalculator.ThrottlerName.DurableDataLog);
+
+        int cacheDelay = 6000;
+        int batchingDelay = 5000;
+        int durableDataLogDelay = 3000;
+
+        cacheCalculator.setDelayMillis(cacheDelay);
+        batchingCalculator.setDelayMillis(batchingDelay);
+        durableDataLogCalculator.setDelayMillis(durableDataLogDelay);
+
+        Consumer<Integer> recordDelay = delayMillis -> { };
+        ArrayList<ThrottlerCalculator.Throttler> throttlers = new ArrayList<>(Arrays.asList(cacheCalculator, batchingCalculator, durableDataLogCalculator));
+
+        @Cleanup
+        TestThrottler t = new TestThrottler(this.containerId, wrap(throttlers), executorService(), metrics, recordDelay);
+
+        val t1 = t.throttle();
+        Assert.assertFalse("Not expected throttle future to be completed yet.", t1.isDone());
+
+        int throttlerActiveTime = 2000;
+        // Suppose a second goes by.
+        TimeUnit.MILLISECONDS.sleep(throttlerActiveTime);
+        // Now the Cache state has changed.
+        // Change delay of Cache throttler in order to make the ThrottlerCalculator pick up the Batching Throttler.
+        cacheCalculator.setDelayMillis(1000);
+        // Change source of throttling.
+        t.notifyThrottleSourceChanged();
+        // Wait for the batching throttler to complete.
+        TestUtils.await(t1::isDone, 5, TIMEOUT_MILLIS);
+        // Throttler should report it spent at least `throttlerActiveTime` throttled.
+        AssertExtensions.assertGreaterThanOrEqual(
+                "Expected to be greater than the time left uninterrupted.",
+                throttlerActiveTime,
+                (int) getThrottlerMetric(ThrottlerCalculator.ThrottlerName.Cache)
+        );
+        // Force an interruption. The throttler should take the min of the left over delay from the previous throttler,
+        // or the throttling delay originally assigned.
+        AssertExtensions.assertLessThanOrEqual(
+                "Expected to be greater than the time left uninterrupted.",
+                 min(batchingDelay, cacheDelay - throttlerActiveTime),
+                (int) getThrottlerMetric(ThrottlerCalculator.ThrottlerName.Batching)
+        );
+
+        durableDataLogDelay = max(cacheDelay, batchingDelay) + 1000;
+        // This should force the throttler calculator to pick up the durable data log throttler.
+        durableDataLogCalculator.setDelayMillis(durableDataLogDelay);
+        val t2 = t.throttle();
+        Assert.assertFalse("Not expected throttle future to be completed yet.", t2.isDone());
+        // Arbitrary amount of time goes by.
+        TimeUnit.MILLISECONDS.sleep(throttlerActiveTime);
```

Comment:
I really do not like sleeping in tests (or anywhere else for that matter). Can you rewrite this test without the need to sleep?